### PR TITLE
Expose expression base class publicly and simplify public AST API

### PR DESCRIPTION
**Pull request by vyasr**
This PR performs the renames `ast::expression`->`ast::operation` and `ast::node`->`ast::expression`, which more accurately reflects that literals and column references are also valid expressions. The new `expression` class is publicly exposed, allowing client code to maintain a list of expression pointers composed of different subclasses of expression, and all APIs now accept arbitrary expressions as input so that it is possible to evaluate e.g. a raw column reference on a table. The old `nodes.hpp` and `operators.hpp` files have been merged into a single `expressions.hpp` header that now contains the entire public API for AST construction, but some implementation details have been moved into an `expressions.cpp` source file so dependencies on detail APIs like the `expression_parser` are restricted to a single TU, which dramatically simplifies recompilation of the code on small changes. Many more parameters are now accepted by (const) ref rather than by value, avoiding unnecessary copies and improving const correctness of the code.

### CODE REVIEW

**Comment by vyasr** [comment]
@jlowe this PR addresses a number of the issues that we have discussed offline (some of which are captured in #8145). Java tests will not pass until I update those bindings (I'm happy to take a stab at it), but I'd like to reach consensus on the C++ changes first.

**Review by robertmaynard — APPROVED** [review]
CMake changes approved

File: cpp/include/cudf/ast/expressions.hpp
Comment (jlowe) [inline cpp/include/cudf/ast/expressions.hpp:249]:
```suggestion
 * @brief An operation expression holds an operator and zero or more operands.
```

**Review by jlowe — COMMENTED** [review]
Overall the changes look OK.

> Java tests will not pass until I update those bindings (I'm happy to take a stab at it)

Should be straightforward, let me know if you want me to do the update.  Note that there is now downstream code that is using the Java APIs for cudf AST.  If the Java APIs make a breaking change in this PR then I'll need to be ready to update that downstream code ASAP once the new cudf jar snapshot is published.

File: cpp/src/ast/expressions.cpp
Comment (ttnghia) [inline cpp/src/ast/expressions.cpp:2]: Not sure if this file is new or renamed. If it is new, please update header year.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/ast/expressions.cpp
Comment (vyasr) [reply cpp/src/ast/expressions.cpp]: Good catch. Updated.

File: java/src/main/java/ai/rapids/cudf/ast/Operation.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ast/Operation.java:24]: Now that all AST nodes can be an expression, this method should move to `AstExpression` because any expression can become a `CompiledExpression`.  I'm not sure this Java class still needs to exist afterwards.

File: java/src/main/native/src/jni_compiled_expr.hpp
Comment (jlowe) [inline java/src/main/native/src/jni_compiled_expr.hpp:42]: Now that the base expression class has a virtual destructor, there's no need to track the different node types separately.  These can all be in a single `std::vector<std::unique_ptr<cudf::ast::expression>>`.  We still need to track the scalars corresponding to AST literals separately, but otherwise everything else can be combined.  I can do this in a followup PR if you'd rather not refactor that here.

**Comment by vyasr** [comment]
rerun tests

**Comment by vyasr** [comment]
@jlowe once this PR is approved would it be best to hold off merging until you've had a chance to update the sparks-rapids consumers of this API?

File: java/src/main/native/src/jni_compiled_expr.hpp
Comment (jlowe) [inline java/src/main/native/src/jni_compiled_expr.hpp:10]: Looks like an errant paste?
```suggestion
 * Unless required by applicable law or agreed to in writing, software
```

File: java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java:45]: This shouldn't be an identity operation anymore, we can use the column reference directly.  This comment applies to all the places where we're using a unary identity operation as a hack around the old restriction that column references and literals couldn't be used directly as expressions.

**Review by jlowe — COMMENTED** [review]
> would it be best to hold off merging until you've had a chance to update the sparks-rapids consumers of this API?

I should have a draft PR to update the RAPIDS Accelerator later today, so I don't think there's a need to wait.

File: java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java
Comment (vyasr) [reply java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java]: Done. In the process of getting this to work I had to modify `compile_serialized_ast` to support literals and column references, and now that they cover all the same types I modified it to just call `compile_expression` directly. Let me know if that causes any problems that I'm not anticipating.

**Review by jlowe — APPROVED** [review]

**Review by dillon-cullinan — APPROVED** [review]

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**
